src/lib-mpi/mpiulfm.h

```diff
@@ -0,0 +1,145 @@
+#ifndef PyMPI_MPIULFM_H
+#define PyMPI_MPIULFM_H
+
+#if MPI_VERSION < 5
+
+#if defined(MPICH_NAME) && (MPICH_NAME >= 3)
+#if MPICH_NUMVERSION >= 30200000
+#define PyMPI_HAVE_MPIX_ERR_REVOKED 1
+#define PyMPI_HAVE_MPIX_ERR_PROC_FAILED 1
+#define PyMPI_HAVE_MPIX_ERR_PROC_FAILED_PENDING 1
+#define PyMPI_HAVE_MPIX_Comm_revoke 1
+#define PyMPI_HAVE_MPIX_Comm_failure_ack 1
+#define PyMPI_HAVE_MPIX_Comm_failure_get_acked 1
+#define PyMPI_HAVE_MPIX_Comm_agree 1
+#define PyMPI_HAVE_MPIX_Comm_shrink 1
+#endif
+#endif
+
+#if defined(OPEN_MPI)
+#include <mpi-ext.h>
+#ifdef OMPI_HAVE_MPI_EXT_FTMPI
+#define PyMPI_HAVE_MPIX_ERR_REVOKED 1
+#define PyMPI_HAVE_MPIX_ERR_PROC_FAILED 1
+#define PyMPI_HAVE_MPIX_ERR_PROC_FAILED_PENDING 1
+#define PyMPI_HAVE_MPIX_Comm_revoke 1
+#define PyMPI_HAVE_MPIX_Comm_is_revoked 1
+#define PyMPI_HAVE_MPIX_Comm_failure_ack 1
+#define PyMPI_HAVE_MPIX_Comm_failure_get_acked 1
+#define PyMPI_HAVE_MPIX_Comm_agree 1
+#define PyMPI_HAVE_MPIX_Comm_iagree 1
+#define PyMPI_HAVE_MPIX_Comm_shrink 1
+#endif
+#endif
+
+#endif
+
+
+#ifndef PyMPI_HAVE_MPI_ERR_REVOKED
+#ifdef  PyMPI_HAVE_MPIX_ERR_REVOKED
+#undef  MPI_ERR_REVOKED
+#define MPI_ERR_REVOKED MPIX_ERR_REVOKED
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_ERR_PROC_FAILED
+#ifdef  PyMPI_HAVE_MPIX_ERR_PROC_FAILED
+#undef  MPI_ERR_PROC_FAILED
+#define MPI_ERR_PROC_FAILED MPIX_ERR_PROC_FAILED
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_ERR_PROC_FAILED_PENDING
+#ifdef  PyMPI_HAVE_MPIX_ERR_PROC_FAILED_PENDING
+#undef  MPI_ERR_PROC_FAILED_PENDING
+#define MPI_ERR_PROC_FAILED_PENDING MPIX_ERR_PROC_FAILED_PENDING
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_Comm_revoke
+#ifdef  PyMPI_HAVE_MPIX_Comm_revoke
+static int PyMPI_Comm_revoke(MPI_Comm comm)
+{
+  return MPIX_Comm_revoke(comm);
+}
+#undef  MPI_Comm_revoke
+#define MPI_Comm_revoke PyMPI_Comm_revoke
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_Comm_is_revoked
+#ifdef  PyMPI_HAVE_MPIX_Comm_is_revoked
+static int PyMPI_Comm_is_revoked(MPI_Comm comm, int *flag)
+{
+  return MPIX_Comm_is_revoked(comm, flag);
+}
+#undef  MPI_Comm_is_revoked
+#define MPI_Comm_is_revoked PyMPI_Comm_is_revoked
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_Comm_get_failed
+#ifdef  PyMPI_HAVE_MPIX_Comm_failure_get_acked
+static int PyMPI_Comm_get_failed(MPI_Comm comm, MPI_Group *group)
+{
+  return MPIX_Comm_failure_get_acked(comm, group);
+}
+#undef  MPI_Comm_get_failed
+#define MPI_Comm_get_failed PyMPI_Comm_get_failed
+#endif
+#endif
+
+#ifndef PyMPI_HAVE_MPI_Comm_ack_failed
+#ifdef  PyMPI_HAVE_MPIX_Comm_failure_ack
+static int PyMPI_Comm_ack_failed(MPI_Comm comm, int num_to_ack, int *num_acked)
```

Comment:
@bosilca I did not quite understand the second paragraph of your comment [here](https://github.com/mpi4py/mpi4py/pull/307#discussion_r1107528400). Please take a look at this piece of code, where I attempt to implement the MPI-5.0 API  with the MPIX API currently available in Open MPI.

This is my implementation of MPI-5.0 `MPI_Comm_ack_failed()` using `MPIX_Comm_failure_ack()` that Open MPI currently provides. As you can see, I check the value of `num_to_ack` on input, and if it is less than the communicator size, I error (because otherwise I cannot satisfy the "ack all failures" semantics). However, note that I'm returning `MPI_UNDEFINED` for `num_acked`, because I don't know how to get the value.  Perhaps I should use `MPIX_Comm_failure_get_acked(comm, &group)` to query the failed group size and set `num_acked`?

The advise to users also say *One may query, without side effect, for the number of currently aknowledged [<- typo is not mine!] process failures in comm by supplying 0 in num_to_ack.*. I'm not sure whether I can support this semantics by querying the failed group with  `MPIX_Comm_failure_get_acked`. In any case, the best we could do is to return num_acked with either zero or failed_group.size.

Reply:
Your implementation of PyMPI_Comm_ack_failed seems correct to me. However, your implementation of PyMPI_Comm_get_failed is not, because `MPIX_Comm_failure_get_acked` only returns the errors you have already acknowledged, when the corresponding MPI function is able to return more failures, even those that have not yet been acknowledged. That being said I don't think you can implement the new API using the old one, and have the same expressivity. 

Btw, https://github.com/open-mpi/ompi/pull/10507 is in the final stages and will bring the new ULFM API to OMPI.

My second comment was about implementing the old API with the new API, not sure how relevant it is in this discussion.

Reply:
> However, your implementation of PyMPI_Comm_get_failed is not, because MPIX_Comm_failure_get_acked only returns the errors you have already acknowledged, when the corresponding MPI function is able to return more failures, even those that have not yet been acknowledged.

Oh, I totally missed that part... You see, I'm still a bit lost following together the original ULFM proposal and the MPI accepted chapter...

Perhaps I should do it the following way?
```c
static int PyMPI_Comm_get_failed(MPI_Comm comm, MPI_Group *group)
{
  MPIX_Comm_failure_ack(comm);
  return MPIX_Comm_failure_get_acked(comm, group);
}
```
What do you think?

> Btw, open-mpi/ompi#10507 is in the final stages and will bring the new ULFM API to OMPI.

What's the expected timeline for that PR to get to v5.0.x? Perhaps I should wait until it is merged to get this one in? In that case, I'll have keep updating your branch, and then ask a final time for to refresh your commit signature, sorry for the noise.
PS: Will the new ULFM API be backported to v4.0.x ?

Reply:
If you do `failure_ack + get_acked` then you automatically ack failed processes on behalf of the user. The bottom line is that the 2 API are not equivalent, it is difficult to implement the new one using the old one. At least not in an easy way. 

The hardway is by having a duplicate communicator that is not visible to the user, and doing what you propose above on the duplicate. This way you can gather information about new failures, but without disturbing the user visible state of the communicator.

The new API PR should be merged later today, and then moved to the next 5.0 rc. The community OMPI 4.0 does not contains ULFM, and I'm not sure we want to backport the new API into our own fork.